text/3678-final.md

```diff
@@ -0,0 +1,184 @@
+same crate or module.
+
+`final fn` cannot be combined with `default fn`.
+
+`final` is only allowed in trait definitions. `final` is not allowed on impls
+or their items, non-trait functions, or `extern` blocks.
+
+A `final fn` has no impact on the `dyn`-compatibility of a trait.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+As with any language feature, this adds more surface area to the language.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Instead of or in addition to this, we could allow inherent `impl` blocks for a
+`Trait` (e.g. `impl Trait { ... }` without `for Type`). People today already
+occasionally write `impl dyn Trait` blocks, since `dyn Trait` is a type and
+supports inherent impl blocks; this change would allow generalizing such blocks
+by deleting the `dyn`. This has the potential for conceptual complexity or
+confusion for new users, as well as potentially affecting the quality of
+diagnostics. (It also used to have a meaning in Rust 2015: the same meaning
+`impl dyn Trait` now has.) However, it would provide orthogonality, and an
+interesting conceptual model.
+
+Rather than using `final`, we could use the `impl(visibility)` syntax from
+[RFC 3323](https://rust-lang.github.io/rfcs/3323-restrictions.html). This would
+allow more flexibility (such as overriding a method within the crate but not
+outside the crate), and would be consistent with other uses of RFC 3323. On the
+other hand, such flexibility would come at the cost of additional complexity.
+We can always add such syntax for the more general cases in the future if
+needed; see the future possibilities section.
+
+Rather than using `final`, we could use `#[final]`. This
+concept is somewhat similar to "final" methods in other languages, and we
+already have the `final` keyword reserved so we could use either an attribute
+or a keyword.
+
+It's possible to work around the lack of this functionality by placing the
+additional methods in an extension trait with a blanket implementation.
+However, this is a user-visible API difference: the user must import the
+extension trait, and use methods from the extension trait rather than from the
+base trait.
+
+# Prior art
+[prior-art]: #prior-art
+
+This feature is similar to `final` methods in Java or C++.
+
+It's also similar to `sealed` in C#, where `sealed class` is something from which you can't derive and a base class can use `sealed` on a method to say derived classes can't `override` it.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None yet.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+We could add the optimization of not putting `final` methods in the trait
+vtable. A design for this should take into account cases where people *do* want
+`final` methods to appear in the vtable, because they benefit substantially
+from monomorphization.
+
```

Comment:
```suggestion

We could allow `final fn` methods on `#[marker]` traits, which are currently not allowed to have any methods (because they can't allow different implementations in different `impl`s).

```